### ISSUE TITLE
cancelDownloadTask before setImage can make some images cancelled to never appear

### ISSUE THREAD

**Issue by damien-nd** [onevcat/Kingfisher#532]
I have try #183 and #421 but it doesn't work for me.

Here is my code:

	kf.cancelDownloadTask()
	kf.setImage(with: url,
		placeholder: placeholderImage,
		options: [.transition(.fade(0.3)), .backgroundDecode])

I do it on UIImageView extension and most of the time it is working great.
But sometimes, it leave imageView with the placeholder (mainly when a lot of images and scrolling fast).
If I add a completionHandler, the error code is -999 for "cancelled".
The normal behavior is to have this cancelled download, but wait a few seconds and the new image will appear.
Sometime, whatever I wait 5 minutes, the image stay on the placeHolder image. If I scroll to put this imageView out of the screen and put it back, the image will show directly with the fade animation.

I can't reproduce it easily and I don't find a nice api to throw a lot of images. Sorry to not help more.

**Comment by onevcat** [comment]
Thanks for reporting this. I will take a deeper look on this.

Before I could conclude what happened, could you try to move the calling on `kf.cancelDownloadTask ` to the `didEndDisplayingCell` method (if you are using this in a cell of a table view or collection view) to see whether the same thing occurs?

**Comment by octopepper** [comment]
I'm sorry but in `didEndDisplayingCell` I still can reproduce.
It is a tableview with 3 images. I should change it for collection view as now the design is the same.
On table view with just 1 image, it is harder to reproduce it, but sometime I face it.

Thank you for investing it.

**Comment by damien-nd** [comment]
I'm sorry but in `didEndDisplayingCell` I still can reproduce.
It is a tableview with 3 images. I should change it for collection view as now the design is the same.
On table view with just 1 image, it is harder to reproduce it, but sometime I face it.

Thank you for investing it.

**Comment by OneSman7** [comment]
Hello!

Using latest version from pod. I can confirm that cancelling image load in similar example does nothing.
My view controller has one image view. In viewDidLoad I write following:

`let urlString = "https://s-media-cache-ak0.pinimg.com/736x/c6/8e/1c/c68e1c54710c1a002d4317917e745337.jpg"` 
`let url = URL(string: urlString)` 
`imageView.kf.setImage(with: url)` 
`imageView.kf.cancelDownloadTask()` 

Then I place breakpoints on 

`imageView.kf.cancelDownloadTask()` 

and `retrieveImage`, `tryToRetrieveImageFromCache` methods of `KingfisherManager`, `cancel` method of `RetrieveImageTask`.

I was able to determine that `imageView.kf.cancelDownloadTask()` is called first, but `cancel` on task is never called. Then image load proceeds as usual.

I see this as a major bug.

**Comment by onevcat** [comment]
@OneSman7 Great catch! 👍 I found an issue on this. Instead of a associated object problem, it is a race condition in cancel task. I will give a fix on it tomorrow!

@damien-nd And your issue might be related to another disk cache cancelling problem, which the callback not called properly. I will also fix it tomorrow since it is now quite late for me and I need a sleep.

**Comment by yuvalt** [comment]
From what I've seen in my code, this can still happen if you pass the option " .cacheMemoryOnly".

**Comment by mono0926** [comment]
@onevcat 

On version 3.9.1, this problem was reproduced.


On demo project, `cancelDownloadTask` rarely called, so I modified the demo source code to reproduce it easily:
https://github.com/mono0926/Kingfisher/tree/fix-missing-image (commit: https://github.com/mono0926/Kingfisher/commit/e2fe4b53f831b745e55b2f6c351f4d94aff2f433 )

1. Call `kf.setImage`(will be cancelled)
2. Call `task.cancel()`
3. Call `kf.setImage`
4. Image should be shown, but sometimes missing.

Add break points to `ImageDownloader.swift`.
<img width="902" alt="screen shot 2017-06-01 at 18 55 18" src="https://cloud.githubusercontent.com/assets/1255062/26674603/f4ac8cf6-46fb-11e7-9752-7dfaee56c7cd.png">

The first break point is to wait for `task.internalTask.cancel()` should be called.
After `task.internalTask.cancel()` called, some images will be missing.

![simulator screen shot jun 1 2017 18 57 39](https://cloud.githubusercontent.com/assets/1255062/26674693/32f3805a-46fc-11e7-8d53-4f48854d7ac9.png)

I'm troubled with it on my project, so I tried to fix, but I couldn't :(
Could you fix it?

**Comment by mono0926** [comment]
> I'm troubled with it on my project

This failes setting an image sometimes:

```swift
kf.cancelDownloadTask()
kf.setImage(with: 🐶)
```

By inserting `DispatchQueue.main.async`, that seems to be fixed 👀 

```swift
kf.cancelDownloadTask()
DispatchQueue.main.async { [weak self] in
    self?.kf.setImage(with: 🐶)
}
```

**Comment by onevcat** [comment]
I am giving it a fix now. :-/

**Comment by onevcat** [comment]
@mono0926 Could you give [this branch](https://github.com/onevcat/Kingfisher/tree/fix/cancel-semaphore) a try to see whether it works for you? Thanks!

**Comment by mono0926** [comment]
@onevcat 

Thanks for the quick fix!
It works fine now 🎉

**Comment by onevcat** [comment]
没问题
Spark by Readdle

**Comment by inamiy** [comment]
Guessing this issue is fixed in #698 :)
